File: aggregate/twitter.py
from datetime import datetime, timezone

import base64
import configparser
import feedparser
import iso8601
import requests
import time

import utility

# Optional features
config = configparser.ConfigParser()
config.read( '../features.ini' )

api = 'http://localhost:3000/api'

# List of Twitter accounts
req = requests.get( api + '/twitter' )
twitters = req.json()

for twitter in twitters:
  # Retrieve user timeline
  print( 'Load: ' + twitter['id'] )
  req = requests.get( api + '/twitter/timeline/' + twitter['screen_name'] )
  timeline = req.json()

  # Update account statistics
  user = timeline[0]['user']

  twitter['name'] = None if len( user['name'] ) == 0 else user['name']
  twitter['image'] = None if len( user['profile_image_url_https'] ) == 0 else user['profile_image_url_https']
  twitter['followers'] = user['followers_count']
  twitter['friends'] = user['friends_count']
  twitter['favorites'] = user['favourites_count']
  twitter['count'] = user['statuses_count']
  twitter['location'] = None if len( user['location'] ) == 0 else user['location']
  twitter['description'] =  None if len( user['description'] ) == 0 else user['description']
  twitter['url'] =  None if len( user['url'] ) == 0 else user['url']

  # Update the database
  req = requests.put( api + '/twitter/' + twitter['id'], json = twitter )
  row = req.json()

  # Look at each entry
  for status in timeline:    
    # Formalize entity
    record = {
      'twitter_id': twitter['id'],
      'published_at': status['created_at'],
      'status': status['id_str'],
      'link': 'https://twitter.com/' + twitter['screen_name'] + '/status/' + status['id_str'],
      'full_text': status['full_text'],
      'favorite': status['favorite_count'],
      'retweet': status['retweet_count'],
      'hashtags': [],
      'mentions': [],
      'urls': []
    }

    # Check database
    req = requests.get( api + '/twitter/status/id/' + str( record['status'] ) )
    matches = req.json()

    # Does not exist
    if matches == None:
      # Parse Twitter date format (RFC 2822)
      published = datetime.strptime( record['published_at'], '%a %b %d %H:%M:%S +%f %Y' )      
      record['published_at'] = published.strftime( '%Y-%m-%d %H:%M:%S' )

      # Populate hashtags
      # Array of objects
      # Not a simple join
      for hashtag in status['entities']['hashtags']:
        record['hashtags'].append( hashtag['text'] )

      # Populate mentions
      for mention in status['entities']['user_mentions']:
        record['mentions'].append( mention['screen_name'] )

      # Populate URLs
      for link in status['entities']['urls']:
        record['urls'].append( link['expanded_url'] )        

      # Create status
      req = requests.post( api + '/twitter/status', json = record )
      insert = req.json()

      # Media
      # Need status ID (primary key from insert) first
      # Media key from Twitter is not always present
      if 'media' in status['entities']:
        for media in status['entities']['media']:
          # Does file exist in database
          encoded = base64.urlsafe_b64encode( media['media_url_https'].encode( 'utf-8' ) )
          req = requests.get( api + '/media/url/' + str( encoded, 'utf-8' ) )        
          matches = req.json()          

          # Nope
          if matches == None:
            record = {
              'url': media['media_url_https'],
              'keywords': None
            }

            # Analyze image
            # Optional feature
            if config['WATSON'].getboolean( 'Vision' ) == True:
              encoded = base64.urlsafe_b64encode( record['url'].encode( 'utf-8' ) )           
              req = requests.get( api + '/watson/vision/' + str( encoded, 'utf-8' ) )
              record['keywords'] = req.json()            

            # Create media record
            req = requests.post( api + '/media', json = record )
            matches = req.json()

          # Associate with status
          req = requests.post( api + '/twitter/status/' + insert['id'] + '/media', json = {
            'media_id': matches['id']
          } )
          associate = req.json()

          print( 'Pict: ' + associate['id'] )

      print( 'Make: ' + insert['id'] )
    else:
      matches['favorite'] = record['favorite']
      matches['retweet'] = record['retweet']

      # Update database
      # Do not care about date range (like other feed sources)
      # Data is already provided in timeline request
      # No additional API usage calls are needed
      # In that case - update all the things
      # Items will stop getting updated once they fall off users timeline
      req = requests.put( api + '/twitter/status/' + matches['id'], json = matches )
      
      print( 'Edit: ' + matches['id'] )

File: routes/twitter.js
const express = require( 'express' );
const rp = require( 'request-promise-native' );
const uuidv4 = require( 'uuid' );

// Router
let router = express.Router();

// Test
router.get( '/test', ( req, res ) => {    
  res.json( {twitter: 'Test'} );
} );

// Read timeline for a given user
// Part of API surface due to authentication
router.get( '/timeline/:screen_name', async ( req, res ) => {
  let timeline = await rp( {
    url: 'https://api.twitter.com/1.1/statuses/user_timeline.json',
    method: 'GET',
    headers: {
      Authorization: `${req.twitter.token_type} ${req.twitter.access_token}`
    },
    qs: {
      screen_name: req.params.screen_name,
      count: 100,
      tweet_mode: 'extended'
    },
    json: true      
  } );

  res.json( timeline );
} );

// Read single Twitter account by ID
router.get( '/:id', ( req, res ) => {
  let twitter = req.db.prepare( `
    SELECT
      Twitter.uuid AS "id",
      Twitter.created_at, 
      Twitter.updated_at,
      Developer.uuid AS "developer_id",
      Twitter.user,
      Twitter.joined_at,
      Twitter.name,
      Twitter.screen_name,
      Twitter.image,
      Twitter.followers,
      Twitter.friends,
      Twitter.favorites,
      Twitter.count,
      Twitter.location,
      Twitter.description,
      Twitter.url
    FROM 
      Developer, 
      Twitter
    WHERE 
      Twitter.developer_id = Developer.id AND
      Twitter.uuid = ?
  ` )
  .get( 
    req.params.id 
  );

  if( twitter === undefined ) {
    twitter = null;
  }

  res.json( twitter );
} );

// Read all Twitter accounts
router.get( '/', ( req, res ) => {
  let twitters = req.db.prepare( `
    SELECT
      Twitter.uuid AS "id",
      Twitter.created_at, 
      Twitter.updated_at,
      Developer.uuid AS "developer_id",
      Twitter.user,
      Twitter.joined_at,
      Twitter.name,
      Twitter.screen_name,
      Twitter.image,
      Twitter.followers,
      Twitter.friends,
      Twitter.favorites,
      Twitter.count,
      Twitter.location,
      Twitter.description,
      Twitter.url
    FROM 
      Developer,
      Twitter
    WHERE Twitter.developer_id = Developer.id
    ORDER BY datetime( Twitter.updated_at ) DESC
  ` )
  .all();

  res.json( twitters );
} );

// Create
router.post( '/', async ( req, res ) => {
  let record = {
    id: null,
    uuid: uuidv4(),
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
    developer_uuid: req.body.developer_id,
    user: 0,
    joined_at: null,
    name: null,
    screen_name: req.body.screen_name,
    image: null,
    followers: 0,
    friends: 0,
    favorites: 0,
    count: 0,
    location: null,
    description: null,
    url: null
  };  

  let existing = req.db.prepare( `
    SELECT
      Twitter.uuid AS "id",
      Twitter.created_at,
      Twitter.updated_at,
      Developer.uuid AS "developer_id",
      Twitter.user,
      Twitter.joined_at,
      Twitter.name,
      Twitter.screen_name,
      Twitter.image,
      Twitter.followers,
      Twitter.friends,
      Twitter.favorites,
      Twitter.count,
      Twitter.location,
      Twitter.description,
      Twitter.url
    FROM
      Developer,
      Twitter
    WHERE
      Twitter.developer_id = Developer.id AND
      Twitter.screen_name = ?
  ` ).get(
    record.screen_name
  );

  if( existing === undefined ) {
    let profile = await rp( {
      url: 'https://api.twitter.com/1.1/users/show.json',
      method: 'GET',
      headers: {
        Authorization: `${req.twitter.token_type} ${req.twitter.access_token}`
      },
      qs: {
        screen_name: record.screen_name
      },
      json: true
    } );
  
    record.user = profile.id;
    record.joined_at = new Date( profile.created_at ).toISOString();
    record.name = profile.name;
    record.image = profile.profile_image_url_https;
    record.followers = profile.followers_count;
    record.friends = profile.friends_count;
    record.favorites = profile.favourites_count;
    record.count = profile.statuses_count;
    record.location = profile.location;
    record.description = profile.description;
    record.url = profile.url;
  
    let developer = req.db.prepare( `
      SELECT Developer.id
      FROM Developer
      WHERE Developer.uuid = ?
    ` )
    .get( 
      record.developer_uuid
    );
    record.developer_id = developer.id;
  
    let info = req.db.prepare( `
      INSERT INTO Twitter
      VALUES ( ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ? )
    ` )
    .run(
      record.id,
      record.uuid,
      record.created_at,
      record.updated_at,
      record.developer_id,
      record.user,
      record.joined_at,
      record.name,
      record.screen_name,
      record.image,
      record.followers,
      record.friends,
      record.favorites,
      record.count,
      record.location,
      record.description,
      record.url
    );
  
    record = {
      id: record.uuid,
      created_at: record.created_at,
      updated_at: record.updated_at,
      developer_id: record.developer_uuid,
      user: record.user,
      joined_at: record.joined_at,
      name: record.name,
      screen_name: record.screen_name,
      image: record.image,
      followers: record.followers,
      friends: record.friends,
      favorites: record.favorites,
      count: record.count,
      location: record.location,
      description: record.description,
      url: record.url
    };
  } else {
    record = existing;
  }

  res.json( record );
} );

// Update
router.put( '/:id', ( req, res ) => {
  let record = {
    uuid: req.params.id,
    updated_at: new Date().toISOString(),
    developer_uuid: req.body.developer_id,
    user: req.body.user,
    joined_at: req.body.joined_at,
    name: req.body.name,
    screen_name: req.body.screen_name,
    image: req.body.image,
    followers: req.body.followers,
    friends: req.body.friends,
    favorites: req.body.favorites,
    count: req.body.count,
    location: req.body.location,
    description: req.body.description,
    url: req.body.url
  };

  let developer = req.db.prepare( `
    SELECT Developer.id
    FROM Developer
    WHERE Developer.uuid = ?
  ` )
  .get( 
    record.developer_uuid
  );
  record.developer_id = developer.id;

  let info = req.db.prepare( `
    UPDATE Twitter
    SET 
      updated_at = ?,
      developer_id = ?,
      user = ?,
      joined_at = ?,
      name = ?,
      screen_name = ?,
      image = ?,
      followers = ?,
      friends = ?,
      favorites = ?,
      count = ?,
      location = ?,
      description = ?,
      url = ?
    WHERE uuid = ?
  ` )
  .run(
    record.updated_at,
    record.developer_id,
    record.user,
    record.joined_at,
    record.name,
    record.screen_name,
    record.image,
    record.followers,
    record.friends,
    record.favorites,
    record.count,
    record.location,
    record.description,
    record.url,
    record.uuid
  );

  record = req.db.prepare( `
    SELECT
      Twitter.uuid AS "id",
      Twitter.created_at, 
      Twitter.updated_at,
      Developer.uuid AS "developer_id",
      Twitter.user,
      Twitter.joined_at,
      Twitter.name,
      Twitter.screen_name,
      Twitter.image,
      Twitter.followers,
      Twitter.friends,
      Twitter.favorites,
      Twitter.count,
      Twitter.location,
      Twitter.description,
      Twitter.url
    FROM 
      Developer, 
      Twitter
    WHERE 
      Twitter.developer_id = Developer.id AND
      Twitter.uuid = ?
  ` )
  .get( 
    record.uuid
  );

  res.json( record );  
} );

// Delete
router.delete( '/:id', ( req, res ) => {
  let info = req.db.prepare( `
    DELETE FROM Twitter
    WHERE Twitter.uuid = ?
  ` )
  .run(
    req.params.id
  );  

  res.json( {
    id: req.params.id
  } );
} );

// Export
module.exports = router;
